module/zfs/zap_micro.c

```diff
@@ -1202,6 +1239,10 @@ zap_add_impl(zap_t *zap, const char *key,
 		if (mze != NULL) {
 			err = SET_ERROR(EEXIST);
 		} else {
+			if (!mze_canfit_fzap_leaf(zn, zn->zn_hash)) {
+				zap_unlockdir(zap, tag);
+				return (ENOSPC);
```

Comment:
Did you consider upgrading to a fat zap in this case?  That would mitigate the slightly-smaller-than-necessary limit on collisions in the microzap due to your overestimation of the name lengths.

Reply:
I think that makes sense. Will fix it.